File: mobile-apps/alloy-android/app/src/main/kotlin/com/alloyremote/alloy/homeassistant/HomeAssistantDiscovery.kt
package com.alloyremote.alloy.homeassistant

import android.content.Context
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import com.alloyremote.alloy.model.DiscoveredAvr
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch

/**
 * Two-phase HA discovery: find a HA host on the LAN via mDNS, then
 * (once the user has authenticated) enumerate media_player entities
 * that look like AV receivers via the REST API.
 */
class HomeAssistantDiscovery(private val context: Context) {

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private val _hosts = MutableSharedFlow<String>(extraBufferCapacity = 8)
    val hosts: SharedFlow<String> = _hosts.asSharedFlow()

    private val nsd: NsdManager by lazy {
        context.getSystemService(Context.NSD_SERVICE) as NsdManager
    }
    private var listener: NsdManager.DiscoveryListener? = null

    fun startHostDiscovery() {
        val listener = object : NsdManager.DiscoveryListener {
            override fun onStartDiscoveryFailed(p: String?, err: Int) {}
            override fun onStopDiscoveryFailed(p: String?, err: Int) {}
            override fun onDiscoveryStarted(p: String?) {}
            override fun onDiscoveryStopped(p: String?) {}
            override fun onServiceFound(info: NsdServiceInfo) {
                val name = info.serviceName ?: return
                val candidate = "http://$name.local:8123"
                scope.launch { _hosts.emit(candidate) }
            }
            override fun onServiceLost(info: NsdServiceInfo) {}
        }
        runCatching {
            nsd.discoverServices(
                "_home-assistant._tcp",
                NsdManager.PROTOCOL_DNS_SD,
                listener
            )
            this.listener = listener
        }
    }

    fun stopHostDiscovery() {
        listener?.let { runCatching { nsd.stopServiceDiscovery(it) } }
        listener = null
    }

    /**
     * Enumerate AVR-like media_player entities. Must be called after
     * the user has authenticated (i.e. we have a working connection).
     */
    suspend fun discoverReceivers(
        connection: HomeAssistantConnection
    ): List<DiscoveredAvr> {
        val states = connection.fetchStates()
        return states.filter { looksLikeAvr(it) }.map { state ->
            DiscoveredAvr(
                id = "ha:${state.entityId}",
                friendlyName = state.friendlyName ?: state.entityId,
                modelName = state.deviceClass,
                host = connection.baseHost,
                port = connection.basePort,
                sources = setOf(DiscoveredAvr.Source.HomeAssistant),
                haEntityId = state.entityId
            )
        }
    }

    private fun looksLikeAvr(state: HAState): Boolean {
        if (!state.entityId.startsWith("media_player.")) return false
        if (state.deviceClass == "receiver") return true
        val haystack = "${state.entityId} ${state.friendlyName ?: ""}".lowercase()
        val vendors = listOf("denon", "marantz", "yamaha", "onkyo", "pioneer", "anthem")
        return vendors.any { it in haystack }
    }
}

File: mobile-apps/alloy-ios/Alloy/HomeAssistant/HomeAssistantDiscovery.swift
import Foundation
import Network

/// Two-phase HA discovery: find a HA host on the LAN via mDNS, then
/// (once the user has authenticated) enumerate `media_player` entities
/// that look like AV receivers via the REST API.
actor HomeAssistantDiscovery {

    private var hostBrowser: NWBrowser?
    private var hostContinuation: AsyncStream<URL>.Continuation?

    nonisolated let hostStream: AsyncStream<URL>

    init() {
        var cont: AsyncStream<URL>.Continuation!
        self.hostStream = AsyncStream { cont = $0 }
        self.hostContinuation = cont
    }

    /// Browse for `_home-assistant._tcp.local.` and yield candidate URLs.
    func startHostDiscovery() async {
        let browser = NWBrowser(
            for: .bonjour(type: "_home-assistant._tcp", domain: nil),
            using: .tcp
        )
        browser.browseResultsChangedHandler = { [weak self] results, _ in
            Task { await self?.handleHost(results: results) }
        }
        browser.start(queue: .global(qos: .userInitiated))
        self.hostBrowser = browser
    }

    func stopHostDiscovery() {
        hostBrowser?.cancel()
        hostBrowser = nil
    }

    private func handleHost(results: Set<NWBrowser.Result>) {
        for result in results {
            guard case let .service(name, _, _, _) = result.endpoint else { continue }
            // Default to http://<name>.local:8123 — users with TLS
            // will enter their URL manually.
            let urlString = "http://\(name).local:8123"
            if let url = URL(string: urlString) {
                hostContinuation?.yield(url)
            }
        }
    }

    /// Enumerate AVR-like media_player entities. Must be called after
    /// the user has authenticated (i.e. we have a working connection).
    func discoverReceivers(
        using connection: HomeAssistantConnection
    ) async throws -> [DiscoveredAvr] {
        let states = try await connection.fetchStates()
        return states
            .filter { looksLikeAvr($0) }
            .map { state in
                DiscoveredAvr(
                    id: "ha:\(state.entity_id)",
                    friendlyName: state.friendlyName ?? state.entity_id,
                    modelName: state.deviceClass,
                    host: connection.baseURL.host ?? "home-assistant",
                    port: connection.baseURL.port ?? 8123,
                    sources: [.homeAssistant],
                    haEntityId: state.entity_id
                )
            }
    }

    private func looksLikeAvr(_ state: HAState) -> Bool {
        guard state.entity_id.hasPrefix("media_player.") else { return false }
        if state.deviceClass == "receiver" { return true }
        // Fall-back heuristic: entity id or friendly name mentions a
        // known AVR vendor. Refine as we see real HA setups.
        let haystack = "\(state.entity_id) \(state.friendlyName ?? "")".lowercased()
        let vendors = ["denon", "marantz", "yamaha", "onkyo", "pioneer", "anthem"]
        return vendors.contains { haystack.contains($0) }
    }
}
